client/src/pages/Challenges.tsx

```diff
@@ -0,0 +1,197 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import { useCallback, useEffect, useMemo, useRef, useState } from "react";
+import ChallengeCard from "../components/ChallengeCard";
+import { Container, Row, Col, Button, Stack, Dropdown, Form, FormCheck} from "react-bootstrap";
+import { ChallengeDetailsShort } from "../types/ChallengeDetailsShort";
+import { ChallengeDifficulties } from "../types/challengeDifficulties";
+
+
+function Challenges() {
+    const challengesData = useRef<ChallengeDetailsShort[]>();
+
+    const [grid, setGrid] = useState([] as JSX.Element[]);
+    const [isLoading, setIsLoading] = useState(true);
+
+    const [sortByDifficulty, setSortByDifficulty] = useState(false);
+    const [filter, setFilter] = useState(0);
+    const [hideComplete, setHideComplete] = useState(false);
+
+
+    /**
+     * Handles the sorting of challenges by difficulty
+     */
+    const handleSortByDifficulty = useCallback((prevChallengesData: ChallengeDetailsShort[]) => {
+        const sortedChallengesData = [...prevChallengesData];
+        if (!sortByDifficulty) {
+            // Sort challengesData in ascending order by difficulty
+            sortedChallengesData.sort((a, b) => a.difficulty - b.difficulty);
+        } else {
+            // Sort challengesData in descending order by difficulty
+            sortedChallengesData.sort((a, b) => b.difficulty - a.difficulty);
+        }
+        return sortedChallengesData;
+    }, [sortByDifficulty]);
+    
+
+    /**
+     * Create the grid of challenges based on the filter
+     * @param challenges the challenges data
+     * @returns the grid of ChallengeCards
+     */
+    const makeGrid = useCallback((): JSX.Element[] => {
+        if (challengesData.current === undefined) {
+            return [];
+        }
+
+        const grid: JSX.Element[] = [];
+        let row: JSX.Element[] = [];
+        let i = 0;
+        for (const challenge of challengesData.current) {
+            //make sure the challenge is not filtered out
+            if(!(filterIncluded(challenge.difficulty, filter) || filter === 0)) {
+                // console.log("Filtering out: " + challenge.difficulty);
+                continue;
+            }
+
+            row.push(
+                <Col lg={4} key={i} className="mb-4">
+                    <ChallengeCard {...challenge} />
+                </Col>
+            );
+            i++;
+            if (i % 3 === 0) {
+                grid.push(<Row sm={1} lg={3} key={i}>{row}</Row>);
+                row = [];
+            }
+        }
+        if (row.length > 0) {
+            grid.push(<Row key={i}>{row}</Row>);
+        }
+        return grid;
+    }, [filter]);
+
+    useEffect(() => {   
```

Comment:
This effect would cause problems. Separate the challenge id and challenge details state and make it only depend on challenge id. Otherwise it would be constantly calling itself.